File: backend/services/consent_service.py
"""
Consent management service.
"""
from typing import List, Dict, Any
from datetime import datetime
import logging

from .firestore_service import firestore_db
from .audit_service import audit_service
from models import AuditAction

logger = logging.getLogger(__name__)


class ConsentService:
    """
    Service for managing user consent records.
    Handles CRUD operations and audit logging for consent changes.
    """
    
    def get_user_consents_with_providers(self, user_id: str) -> List[Dict[str, Any]]:
        """
        Gets all providers with their consent status for a user.
        Returns ALL providers from Firestore.
        
        IMPORTANT: Colorado is an OPT-OUT state, so consent defaults to TRUE.
        Users must explicitly revoke consent to opt out.
        
        Args:
            user_id: User ID
            
        Returns:
            List of all providers with their consent status (defaults to True for opt-out)
        """
        # Get all providers from Firestore
        all_providers = firestore_db.list_providers()
        
        # Get user's existing consent records
        user_consents = firestore_db.get_user_consents(user_id)
        consent_map = {c['provider_id']: c['consented'] for c in user_consents}
        
        # Build result with all providers
        # Default to True (opted-in) since Colorado is opt-out
        result = []
        for provider in all_providers:
            result.append({
                'id': provider['id'],
                'name': provider['name'],
                'address': provider.get('address'),
                'consented': consent_map.get(provider['id'], True)  # Default TRUE for opt-out state
            })
        
        return result
    
    def toggle_consent(
        self, 
        user_id: str, 
        provider_id: str, 
        consented: bool,
        ip_address: str = None
    ) -> Dict[str, Any]:
        """
        Toggles consent for a provider.
        
        Args:
            user_id: User ID
            provider_id: Provider ID
            consented: New consent status
            ip_address: User's IP address
            
        Returns:
            Updated consent record
        """
        # Check if consent exists
        existing = firestore_db.get_user_consent_for_provider(user_id, provider_id)
        
        if existing:
            # Update existing
            firestore_db.update_consent(existing['id'], {'consented': consented})
            consent_id = existing['id']
        else:
            # Create new
            consent_data = {
                'user_id': user_id,
                'provider_id': provider_id,
                'consented': consented,
                'created_at': datetime.utcnow(),
                'updated_at': datetime.utcnow()
            }
            consent_id = firestore_db.create_consent(consent_data)
        
        # Log action
        action = AuditAction.CONSENT_GRANTED if consented else AuditAction.CONSENT_REVOKED
        audit_service.log_action(
            user_id=user_id,
            action=action,
            provider_id=provider_id,
            ip_address=ip_address
        )
        
        logger.info(f"Consent {'granted' if consented else 'revoked'} for user {user_id}, provider {provider_id}")
        
        return {'success': True, 'consent_id': consent_id}
    
    def grant_all_consents(
        self, 
        user_id: str, 
        provider_ids: List[str],
        ip_address: str = None
    ) -> Dict[str, Any]:
        """
        Grants consent to all providers.
        
        Args:
            user_id: User ID
            provider_ids: List of provider IDs
            ip_address: User's IP address
            
        Returns:
            Success status and count
        """
        for provider_id in provider_ids:
            self.toggle_consent(user_id, provider_id, True, ip_address)
        
        logger.info(f"Granted consent to all {len(provider_ids)} providers for user {user_id}")
        return {'success': True, 'count': len(provider_ids)}
    
    def revoke_all_consents(
        self, 
        user_id: str, 
        provider_ids: List[str],
        ip_address: str = None
    ) -> Dict[str, Any]:
        """
        Revokes consent from all providers.
        
        Args:
            user_id: User ID
            provider_ids: List of provider IDs
            ip_address: User's IP address
            
        Returns:
            Success status and count
        """
        for provider_id in provider_ids:
            self.toggle_consent(user_id, provider_id, False, ip_address)
        
        logger.info(f"Revoked consent from all {len(provider_ids)} providers for user {user_id}")
        return {'success': True, 'count': len(provider_ids)}


# Singleton instance
consent_service = ConsentService()

File: frontend/src/services/consent.service.ts
import { apiClient } from './api.client';
import type { ProviderWithConsent } from '../models/Provider';
import type { UpdateConsentRequest, ConsentResponse } from '../models/ConsentRecord';

/**
 * Response from the consents endpoint.
 */
interface ConsentsResponse {
  providers: ProviderWithConsent[];
}

/**
 * Consent management service.
 * Handles fetching and updating user consent preferences.
 */
class ConsentService {
  /**
   * Fetches all providers with their consent status for the current user.
   * 
   * @returns Promise resolving to array of providers with consent status
   * @throws ApiError if the request fails
   */
  async getConsents(): Promise<ProviderWithConsent[]> {
    try {
      const response = await apiClient.get<ConsentsResponse>('/consents');
      return response.providers;
    } catch (error) {
      console.error('Failed to fetch consents:', error);
      throw error;
    }
  }

  /**
   * Updates consent for a specific provider.
   * 
   * @param request - Update consent request data
   * @returns Promise resolving to the consent response
   * @throws ApiError if the update fails
   */
  async updateConsent(request: UpdateConsentRequest): Promise<ConsentResponse> {
    try {
      const response = await apiClient.post<ConsentResponse>('/consents', request);
      return response;
    } catch (error) {
      console.error('Failed to update consent:', error);
      throw error;
    }
  }

  /**
   * Updates consent for multiple providers at once.
   * Useful for "grant all" or "revoke all" operations.
   * 
   * @param providerIds - Array of provider IDs to update
   * @param consented - New consent status
   * @returns Promise resolving to array of consent responses
   */
  async updateMultipleConsents(
    providerIds: string[],
    consented: boolean
  ): Promise<ConsentResponse[]> {
    try {
      const promises = providerIds.map((providerId) =>
        this.updateConsent({ providerId, consented })
      );
      return await Promise.all(promises);
    } catch (error) {
      console.error('Failed to update multiple consents:', error);
      throw error;
    }
  }

  /**
   * Toggles consent for a specific provider.
   * 
   * @param providerId - Provider ID to toggle
   * @param currentStatus - Current consent status
   * @returns Promise resolving to the consent response
   */
  async toggleConsent(
    providerId: string,
    currentStatus: boolean
  ): Promise<ConsentResponse> {
    try {
      const response = await apiClient.post<ConsentResponse>('/consents/toggle', {
        provider_id: providerId,
        consented: !currentStatus,
      });
      return response;
    } catch (error) {
      console.error('Failed to toggle consent:', error);
      throw error;
    }
  }

  /**
   * Revokes all consents for the current user.
   * 
   * @param providerIds - Array of all provider IDs
   * @returns Promise resolving to the response
   */
  async revokeAllConsents(providerIds: string[]): Promise<ConsentResponse> {
    try {
      const response = await apiClient.post<ConsentResponse>('/consents/revoke-all', {
        provider_ids: providerIds,
      });
      return response;
    } catch (error) {
      console.error('Failed to revoke all consents:', error);
      throw error;
    }
  }

  /**
   * Grants consent to all providers for the current user.
   * 
   * @param providerIds - Array of all provider IDs
   * @returns Promise resolving to the response
   */
  async grantAllConsents(providerIds: string[]): Promise<ConsentResponse> {
    try {
      const response = await apiClient.post<ConsentResponse>('/consents/grant-all', {
        provider_ids: providerIds,
      });
      return response;
    } catch (error) {
      console.error('Failed to grant all consents:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const consentService = new ConsentService();
